libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -3277,14 +3286,17 @@ instance (Ord a, HasSpec fn a) => HasSpec fn (Set a) where
             $ genFromSpecT elemS `suchThatT` (`Set.notMember` s)
         go (n - 1) (Set.insert e s)
 
-  cardinalTypeSpec _ = TrueSpec
+  cardinalTypeSpec (SetSpec _ (MemberSpec xs) _) = leqSpec (2 ^ sizeOf (nub xs))
+  cardinalTypeSpec (SetSpec _ _ _) = TrueSpec
 
   shrinkWithTypeSpec (SetSpec _ es _) as = map Set.fromList $ shrinkList (shrinkWithSpec es) (Set.toList as)
 
   toPreds s (SetSpec m es size) =
-    (Assert ["Subset of " ++ show m] $ subset_ (Lit m) s)
-      <> forAll s (\e -> satisfies e es)
-      <> satisfies (size_ s) size
+    fold $
+      [Assert [show m ++ " is a subset of the set."] $ subset_ (Lit m) s | not $ Set.null m]
+        ++ [ forAll s (\e -> satisfies e es)
+           , satisfies (size_ s) size
+           ]
```

Comment:
Add a comment like
-- Use list comprehension to add the subset_ predicate, only if m is not  Set.empty